File: src/main/java/com/github/kuro46/embedscript/script/processor/GivePermissionProcessor.java
package com.github.kuro46.embedscript.script.processor;

import com.github.kuro46.embedscript.Configuration;
import com.github.kuro46.embedscript.script.ParseException;
import com.google.common.collect.ImmutableList;
import org.apache.commons.lang.ArrayUtils;
import org.bukkit.entity.Player;
import org.bukkit.permissions.PermissionAttachment;
import org.bukkit.plugin.Plugin;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class GivePermissionProcessor implements Processor {
    private final Executor executor;
    private final Parser parser;

    public GivePermissionProcessor(Plugin plugin, Configuration configuration) {
        this.executor = new GivePermissionExecutor(plugin);
        this.parser = new GivePermissionParser(configuration);
    }

    @Override
    public String getKey() {
        return "give-permission";
    }

    @Override
    public String getOmittedKey() {
        return "gp";
    }

    @Override
    public Parser getParser() {
        return parser;
    }

    @Override
    public Executor getExecutor() {
        return executor;
    }

    private static class GivePermissionExecutor extends AbstractExecutor {
        private final Map<Player, PermissionAttachment> attachments = new HashMap<>();
        private final Plugin plugin;

        public GivePermissionExecutor(Plugin plugin) {
            this.plugin = plugin;
        }

        @Override
        public void prepareExecute(Player trigger, ImmutableList<String> matchedValues) {
            PermissionAttachment attachment = matchedValues.isEmpty() ? null : trigger.addAttachment(plugin);
            for (String matchedValue : matchedValues) {
                if (trigger.hasPermission(matchedValue)) {
                    continue;
                }
                attachment.setPermission(matchedValue, true);
            }
            attachments.put(trigger, attachment);
        }

        @Override
        public void endExecute(Player trigger, ImmutableList<String> matchedValues) {
            PermissionAttachment attachment = attachments.get(trigger);
            if (attachment != null) {
                attachment.remove();
            }
        }
    }

    private static class GivePermissionParser extends AbstractParser {
        private final Configuration configuration;

        public GivePermissionParser(Configuration configuration) {
            this.configuration = configuration;
        }

        @Override
        public void build(ScriptBuilder builder, String key, ImmutableList<String> matchedValues) throws ParseException {
            if (!matchedValues.isEmpty()) {
                builder.getScript().putAll(key, matchedValues);
            } else {
                Set<String> preferPermissions = new HashSet<>();
                Map<String, List<String>> permissionsForActions = configuration.getPermissionsForActions();
                for (String action : builder.getScript().values()) {
                    List<String> permissionsForAction = permissionsForActions.get(action);

                    if (permissionsForAction == null) {
                        int skipElement = 1;
                        while (permissionsForAction == null) {
                            String[] split = action.split(" ");
                            ArrayUtils.reverse(split);
                            String[] skipped = Arrays.stream(split)
                                .skip(skipElement)
                                .toArray(String[]::new);
                            if (skipped.length == 0) {
                                break;
                            }
                            ArrayUtils.reverse(skipped);
                            permissionsForAction = permissionsForActions.get(String.join(" ", skipped));

                            skipElement++;
                        }
                    }

                    if (permissionsForAction == null) {
                        continue;
                    }
                    preferPermissions.addAll(permissionsForAction);
                }
                builder.getScript().putAll(key, preferPermissions);
            }
        }
    }
}

File: src/main/kotlin/com/github/kuro46/embedscript/script/processor/GivePermissionProcessor.kt
package com.github.kuro46.embedscript.script.processor

import com.github.kuro46.embedscript.Configuration
import com.github.kuro46.embedscript.script.processor.executor.AbstractExecutor
import com.github.kuro46.embedscript.script.processor.executor.ChildExecutor
import com.github.kuro46.embedscript.script.processor.parser.AbstractParser
import com.github.kuro46.embedscript.script.processor.parser.ChildParser
import org.bukkit.entity.Player
import org.bukkit.permissions.PermissionAttachment
import org.bukkit.plugin.Plugin
import java.util.HashMap
import java.util.HashSet
import java.util.stream.Collectors

class GivePermissionProcessor(plugin: Plugin, configuration: Configuration) : Processor {
    override val executor: ChildExecutor = GivePermissionExecutor(plugin)
    override val parser: ChildParser = GivePermissionParser(configuration)

    override val key = "give-permission"

    override val omittedKey = "gp"

    private class GivePermissionExecutor(private val plugin: Plugin) : AbstractExecutor() {
        private val attachments = HashMap<Player, PermissionAttachment>()

        override fun prepareExecute(trigger: Player, matchedValues: List<String>) {
            val attachment = if (matchedValues.isEmpty()) null else trigger.addAttachment(plugin)
            for (matchedValue in matchedValues) {
                if (trigger.hasPermission(matchedValue)) {
                    continue
                }
                attachment!!.setPermission(matchedValue, true)
            }

            if (attachment != null) {
                attachments[trigger] = attachment
            }
        }

        override fun endExecute(trigger: Player, matchedValues: List<String>) {
            attachments[trigger]?.remove()
        }
    }

    private class GivePermissionParser(private val configuration: Configuration) : AbstractParser() {
        override fun build(builder: ScriptBuilder, key: String, matchedValues: List<String>) {
            if (!matchedValues.isEmpty()) {
                builder.script.putAll(key, matchedValues)
            } else {
                val preferPermissions = HashSet<String>()
                val permissionsForActions = configuration.permissionsForActions
                for (action in builder.script.values()) {
                    var permissionsForAction: List<String>? = permissionsForActions!![action]

                    if (permissionsForAction == null) {
                        var skipElement = 1
                        while (permissionsForAction == null) {
                            val split = action.split(' ')
                                    .dropLastWhile { it.isEmpty() }
                                    .toMutableList()
                            split.reverse()
                            val skipped = split.stream()
                                    .skip(skipElement.toLong())
                                    .collect(Collectors.toList())
                            if (skipped.isEmpty()) {
                                break
                            }
                            skipped.reverse()
                            permissionsForAction = permissionsForActions[skipped.joinToString(" ")]

                            skipElement++
                        }
                    }

                    if (permissionsForAction == null) {
                        continue
                    }
                    preferPermissions.addAll(permissionsForAction)
                }
                builder.script.putAll(key, preferPermissions)
            }
        }
    }
}
